stellargraph/layer/graphsage.py

```diff
@@ -541,54 +482,52 @@ def __call__(self, x: List):
             Output tensor
         """
 
-        def compose_layers(_x: List, layer: int):
+        def apply_layer(x: List, layer: int):
             """
-            Function to recursively compose aggregation layers. When current layer is at final layer, then length of _x
-            should be 1, and compose_layers(_x, layer) returns _x[0].
+            Compute the list of output tensors for a single HinSAGE layer
 
             Args:
-                _x:       List of feature matrix tensors
-                layer:   Current layer index
+                agg (Dict[str, Layer]): Dict of node type to aggregator layer
+                x (List[Tensor]): Inputs to the layer
+                layer (int): Layer index to construct
 
             Returns:
-                _x computed from current layer to output layer
+                Outputs of applying the aggregators as a list of Tensors
+
             """
+            layer_out = []
+            for i in range(self.n_layers - layer):
+                head_shape = x[i].shape[1]
+                # Reshape neighbours per node per layer
+                neigh_in = Reshape((head_shape, self.n_samples[i], self.dims[layer]))(
+                    x[i + 1]
+                )
 
-            def x_next(agg):
-                """
-                Compute the list of tensors for the next layer
-
-                Args:
-                    agg (Layer): Aggregator layer to apply
-
-                Returns:
-                    Output list of tensors of applying the aggregator to inputs
-
-                """
-                return [
-                    agg(
-                        [
-                            Dropout(self.dropout)(_x[i]),
-                            Dropout(self.dropout)(
-                                self._neigh_reshape[layer][i](_x[i + 1])
-                            ),
-                        ]
-                    )
-                    for i in range(self.n_layers - layer)
-                ]
-
-            return (
-                compose_layers(x_next(self._aggs[layer]), layer + 1)
-                if layer < self.n_layers
-                else _x[0]
-            )
+                # Apply aggregator to head node and neighbour nodes
+                layer_out.append(self._aggs[layer]([x[i], neigh_in]))
+
+            return layer_out
 
-        assert isinstance(x, list), "Input features must be a list"
-        assert (
-            len(x) == self.n_layers + 1 > 1
-        ), "Length of input features should match the number of GraphSAGE layers"
+        if not isinstance(xin, list):
+            raise TypeError("Input features to GraphSAGE must be a list")
 
-        return self._normalization(compose_layers(x, 0))
+        if len(xin) != self.n_layers + 1:
```

Comment:
Should the message for the ValueError exception say that '...the number of GraphSAGE layers + 1'?

Reply:
True ... I've changed the error message to reflect this.